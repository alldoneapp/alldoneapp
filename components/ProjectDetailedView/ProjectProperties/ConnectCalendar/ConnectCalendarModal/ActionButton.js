import React from 'react'
import { useSelector } from 'react-redux'
import moment from 'moment'

import Button from '../../../../UIControls/Button'
import { translate } from '../../../../../i18n/TranslationService'
import Backend from '../../../../../utils/BackendBridge'
import GooleApi from '../../../../../apis/google/GooleApi'
import { runHttpsCallableFunction } from '../../../../../utils/backends/firestore'
import { disableOtherProjects, isSomethingConnected } from '../../../../../apis/google/ApiHelper'

export default function ActionButton({ projectId, isConnected, isSignedIn, closePopover, setIsSignedIn }) {
    const loggedUserId = useSelector(state => state.loggedUser.uid)

    const isConnectedAndSignedIn = isConnected && isSignedIn

    const loadEvents = () => {
        GooleApi.listTodayEvents(30).then(({ result }) => {
            runHttpsCallableFunction('addCalendarEventsToTasksSecondGen', {
                events: result?.items,
                projectId,
                uid: loggedUserId,
                email: result.summary,
            })
        })
    }

    const removeOpenEvents = () => {
        runHttpsCallableFunction('removeOldCalendarTasksSecondGen', {
            uid: loggedUserId,
            dateFormated: moment().format('DDMMYYYY'),
            events: [],
            removeFromAllDates: true,
        })
    }

    const onPress = () => {
        !isSomethingConnected() && GooleApi.handleSignOutClick()
        if (isSignedIn && isConnected) {
            disconnect()
        } else {
            GooleApi.handleAuthClick().then(connect)
        }
    }

    const disconnect = () => {
        Backend.getDb()
            .doc(`users/${loggedUserId}`)
            .set({ apisConnected: { [projectId]: { calendar: false } } }, { merge: true })
            .then(removeOpenEvents)
        closePopover()
    }

    const connect = () => {
        Backend.getDb()
            .doc(`users/${loggedUserId}`)
            .set({ apisConnected: { [projectId]: { calendar: true } } }, { merge: true })
            .then(loadEvents)
        closePopover()
        setIsSignedIn(GooleApi.checkAccessGranted())
        disableOtherProjects(projectId)
    }

    return (
        <Button
            title={translate(isConnectedAndSignedIn ? 'Disconnect' : 'Connect')}
            icon={isConnectedAndSignedIn ? 'unlink' : 'link'}
            buttonStyle={{ alignSelf: 'center' }}
            onPress={onPress}
        />
    )
}
