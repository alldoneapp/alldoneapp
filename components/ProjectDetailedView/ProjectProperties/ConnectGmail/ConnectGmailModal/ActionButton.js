import React from 'react'
import { useSelector } from 'react-redux'

import Button from '../../../../UIControls/Button'
import { translate } from '../../../../../i18n/TranslationService'
import Backend from '../../../../../utils/BackendBridge'
import GooleApi from '../../../../../apis/google/GooleApi'
import { connectToGmail } from '../../../../../utils/backends/firestore'
import { disableOtherProjects, isSomethingConnected } from '../../../../../apis/google/ApiHelper'

export default function ActionButton({ projectId, isConnected, isSignedIn, closePopover, setIsSignedIn }) {
    const email = useSelector(state => state.loggedUser.email)
    const loggedUserId = useSelector(state => state.loggedUser.uid)

    const isConnectedAndSignedIn = isConnected && isSignedIn

    const loadEvents = () => {
        GooleApi.listGmail().then(result => {
            connectToGmail({
                projectId,
                date: Date.now(),
                uid: loggedUserId,
                unreadMails: result.threadsTotal,
                email,
            })
        })
    }

    const disconnect = () => {
        Backend.getDb()
            .doc(`users/${loggedUserId}`)
            .set({ apisConnected: { [projectId]: { gmail: false } } }, { merge: true })
        closePopover()
    }

    const connect = () => {
        Backend.getDb()
            .doc(`users/${loggedUserId}`)
            .set({ apisConnected: { [projectId]: { gmail: true } } }, { merge: true })
            .then(loadEvents)
        closePopover()
        setIsSignedIn(GooleApi.checkGmailAccessGranted())
        disableOtherProjects(projectId)
    }

    const onPress = () => {
        !isSomethingConnected() && GooleApi.handleSignOutClick()
        if (isSignedIn && isConnected) {
            disconnect()
        } else {
            GooleApi.handleGmailAuthClick().then(connect)
        }
    }

    return (
        <Button
            title={translate(isConnectedAndSignedIn ? 'Disconnect' : 'Connect')}
            icon={isConnectedAndSignedIn ? 'unlink' : 'link'}
            buttonStyle={{ alignSelf: 'center' }}
            onPress={onPress}
        />
    )
}
