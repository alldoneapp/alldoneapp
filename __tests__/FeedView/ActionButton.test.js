/**
 * @jest-environment jsdom
 */

import React from 'react'
import { toggleSmallScreen } from '../../redux/actions'
import store from '../../redux/store'
import ActionButton from '../../components/FeedView/ActionButton'

import renderer from 'react-test-renderer'
import { render, fireEvent } from 'react-native-testing-library';

describe('ActionButton component', () => {
    it('should render correctly', () => {
        const json = renderer.create(<ActionButton icon={{ icon: null }} text={{ text: null }} />).toJSON()
        expect(json).toMatchSnapshot()
    })

    it('should unmount correctly', () => {
        const tree = renderer.create(<ActionButton icon={{ icon: null }} text={{ text: null }} />)
        tree.unmount()
    })

    it('should updateState correctly', () => {
        // Given
        const expectedValue = true
        store.dispatch(toggleSmallScreen(expectedValue))
        const tree = renderer.create(<ActionButton icon={{ icon: null }} text={{ text: null }} />)
        const instance = tree.getInstance()
        // When
        instance.updateState()
        // Then
        expect(instance.state.smallScreen).toEqual(expectedValue)
    })

    it('should invoke onPress correctly', () => {
        // Given
        const mockFn = jest.fn()
        const { getByTestId } = render(
            <ActionButton icon={{ icon: null }} text={{ text: null }} onPress={mockFn} />
        );
        const touchableOpacity = getByTestId('touchableOpacity')
        // When
        fireEvent.press(touchableOpacity)
        // Then
        expect(mockFn).toBeCalledTimes(1)
    })

    it('should not invoke onPress', () => {
        // Given
        const mockFn = jest.fn()
        const { getByTestId } = render(
            <ActionButton icon={{ icon: null }} text={{ text: null }}/>
        );
        const touchableOpacity = getByTestId('touchableOpacity')
        // When
        fireEvent.press(touchableOpacity)
        // Then
        expect(mockFn).toBeCalledTimes(0)
    })

    it('should hide the Popover', () => {
        // Given
        const tree = renderer.create(<ActionButton icon={{ icon: null }} text={{ text: null }} />)
        const instance = tree.getInstance()
        instance.state.visiblePopover = true
        // When
        instance.hidePopover()
        // Then
        setTimeout(()=> {
            expect(instance.state.visiblePopover).toEqual(false)
        }, 1000)
    })
})
