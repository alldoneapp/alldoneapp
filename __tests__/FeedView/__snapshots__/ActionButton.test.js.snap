// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButton component should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "marginRight": 4,
      },
      null,
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 4,
        "flexDirection": "row",
        "height": 40,
        "opacity": 1,
        "paddingHorizontal": 16,
      }
    }
    testID="touchableOpacity"
  >
    <View>
      <Text
        style={
          Object {
            "color": "#8A94A6",
            "fontFamily": "alldone",
            "fontSize": 24,
            "opacity": 1,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "marginLeft": 12,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Roboto-Medium",
              "fontSize": 14,
              "fontStyle": "normal",
              "letterSpacing": 0.8,
              "lineHeight": 14,
            },
            Object {
              "color": "#8A94A6",
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
